### ISSUE TITLE
The new uploaded pretrained model doesn't seem to be performing very well

### ISSUE THREAD

**Issue by HospitableHost** [garvita-tiwari/PoseNDF#22]
I tested on plausible poses and noise-added poses, and found that the output distance is not very good at distinguishing noisy poses from reasonable poses. Noise-added poses are processed by self.sigma = [0.01, 0.05, 0.1, 0.25, 0.5] and self.sample_distribution =  np.array([0.2, 0.2, 0.2, 0.2, 0.2]). Below are some predictions.
Plausible poses from AMASS trainset:
![image](https://user-images.githubusercontent.com/44576391/221590748-06e59ac7-74b3-4a53-9804-3f6a85786606.png)
Noise-added poses:
![image](https://user-images.githubusercontent.com/44576391/221591615-d74a490d-9b73-48b9-a236-17728fa17b62.png)

**Comment by garvita-tiwari** [comment]
Please share the pose file, so that i can test

**Comment by garvita-tiwari** [comment]
I suspect the following reasons:
1. version 2 of PoseNDF works on positive quaternions, so you need to flip the quaternion to the positive hemisphere, before calculating the distance

Please refer to: https://github.com/garvita-tiwari/PoseNDF/blob/version2/experiments/sample_poses.py#L73, for calculating distance and projecting pose on the manifold.

I have added some examples here:
bad pose(initial, projected pose):
initial distance: tensor(0.0308, device='cuda:0'). 

![init_0002](https://user-images.githubusercontent.com/25167952/221600185-b209e4af-b553-4715-823b-9f8d6a612b06.jpg)![out_0002](https://user-images.githubusercontent.com/25167952/221600272-ab9364cd-4b5d-43e3-b779-f2eea8a627fb.jpg)

good pose(initial, projected pose): -> from ACCAD/Female1General_c3d
initial distance: tensor(0.0005, device='cuda:0')


![init_0006](https://user-images.githubusercontent.com/25167952/221600527-63148da2-e8ce-4fc3-a8c9-5f58a3151b89.jpg)![out_0006](https://user-images.githubusercontent.com/25167952/221600538-497d9968-9fd0-4b41-8bd1-543e3f012654.jpg)

**Comment by HospitableHost** [comment]
Thank you very much! I will compare the difference between the two versions in detail.

**Comment by HospitableHost** [comment]
It seems not good. I tested the pretrained model on the raw pose **directly sampled** from the AMASS trainset, and use the function '**axis_angle_to_quaternion**' and '**quat_flip**' to process the pose, and then input to the model.
But the distance of the plausible poses are still 0.09....

Here is my test code.
```
# test for pretrained model
checkpoint_path = './experiment/2-27.tar'
config_path = './configs/2-27.yaml'
opt = load_config(config_path)
checkpoint = torch.load(checkpoint_path, map_location=opt['train']['device'])
model = PoseNDF(opt)
model.load_state_dict(checkpoint['model_state_dict'])
print(f"checkpoint loaded, its epoch is {checkpoint['epoch']}.")

path = '/mnt/nas_8/datasets/Pose-NDF-raw/ACCAD/Female1General_c3d.npz'
data = np.load(path)
num_poses = 5
rand_idx = np.random.randint(0, len(data['pose_body']), num_poses)
pose = torch.from_numpy(data['pose_body'][rand_idx])[:,:63].reshape(-1,21,3)
pose = axis_angle_to_quaternion(pose)
pose = quat_flip(pose)
distance = model(pose,train = False)
distance = distance[:,0].cpu().detach().clone().numpy()
gt_distance = np.zeros_like(distance)
data = {'predicted distance': distance.tolist(), 'gt distance': gt_distance.tolist(), 'error：': np.absolute(distance-gt_distance).tolist()}
df = pd.DataFrame(data)
print('poses from :{}'.format(path.split('/')[-3:]))
print(df)
```

the dir 'Pose-NDF-raw' refers to the sampled poses processed by 'data/sample_poses.py'

**Comment by HospitableHost** [comment]
Here is some results from the test code.
![image](https://user-images.githubusercontent.com/44576391/221762078-c272ceb5-30c7-4fe3-a72b-7d8a4e6cb2d2.png)

**Comment by garvita-tiwari** [comment]
I checked the same code and I get the following:
{'predicted distance': [0.00030276746838353574, 0.0004039875348098576, 0.00040361887658946216, 0.0008321860223077238, 0.00046724494313821197], 'gt distance': [0.0, 0.0, 0.0, 0.0, 0.0], 'error：': [0.00030276746838353574, 0.0004039875348098576, 0.00040361887658946216, 0.0008321860223077238, 0.00046724494313821197]}

I notice that you get distance using:
distance = model(pose,train = False)

but it should be 
distance = model(pose,train = False)['dist_pred'].

Just to be double sure, I have also re-uploaded the checkpoint.

**Comment by HospitableHost** [comment]
I can't believe that. Why it is still wrong for me using the new one?
https://drive.google.com/file/d/1PjOGLqh4P6RPkjG9Gc7YjFkDmES7YWLa/view?usp=sharing
I have uploaded a npz file, it is raw pose directly from the AMASS' ACCAD.

**Comment by HospitableHost** [comment]
I use your pretrained model, your config file, your model code.
If this data is not wrong, then it must be the problem of the pretrained model.

**Comment by HospitableHost** [comment]
Please help me to test the pose file, I really dont know what is wrong.   o(╥﹏╥)o

**Comment by garvita-tiwari** [comment]
> I can't believe that. Why it is still wrong for me using the new one? https://drive.google.com/file/d/1PjOGLqh4P6RPkjG9Gc7YjFkDmES7YWLa/view?usp=sharing I have uploaded a npz file, it is raw pose directly from the AMASS' ACCAD.

I tested with this file and the code snippet your shared:
{'predicted distance': [0.00072555395308882, 0.0004589210730046034, 0.0005504807340912521, 0.0010260505368933082, 0.0011828476563096046], 'gt distance': [0.0, 0.0, 0.0, 0.0, 0.0], 'error：': [0.00072555395308882, 0.0004589210730046034, 0.0005504807340912521, 0.0010260505368933082, 0.0011828476563096046]}

Did you check this line:
I notice that you get distance using:
distance = model(pose,train = False)

but it should be
distance = model(pose,train = False)['dist_pred'].

**Comment by HospitableHost** [comment]
That's not a problem. I just modify the dict to the distance itself
![image](https://user-images.githubusercontent.com/44576391/221776537-f1367719-d261-41f9-ae8b-03cb15b3f511.png)

**Comment by garvita-tiwari** [comment]
Ok, I think I understood the problem. You are using main branch of the code. But this is for version2 branch of the code.

I think in your code/posendf.py:
pose = torch.nn.functional.normalize(pose.to(device=self.device),dim=1). -> This line is uncommented. (https://github.com/garvita-tiwari/PoseNDF/commit/b53d01502f8c6288b457cc33fe94f4a9aef48213#diff-ae0c20e62851250b817bb51457c32e268c4d57078f02689516a52b049ed27e70)

**Comment by HospitableHost** [comment]
Thank you very much! That is the wrong place.
I tested on the noisy pose, and rendered the noisy pose. It can work well.

**Comment by TimofeyKruk** [comment]
Hi, thank you for your work!
How do you render and draw images? Is there any script available?

**Comment by garvita-tiwari** [comment]
> Hi, thank you for your work! How do you render and draw images? Is there any script available?

You can use this pytorch3d based renderer to create results: https://github.com/garvita-tiwari/PoseNDF/blob/version2/experiments/exp_utils.py#L34

Check the usage here: https://github.com/garvita-tiwari/PoseNDF/blob/version2/experiments/sample_poses.py#L56
